// @cloudinary/transformation-builder-sdk/actions/effect/BackgroundRemoval.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var tslib_es6 = require('../../tslib.es6-f1398b83.cjs');
var internal_Action = require('../../internal/Action.cjs');
var internal_qualifier_QualifierValue = require('../../internal/qualifier/QualifierValue.cjs');
var internal_qualifier_Qualifier = require('../../internal/qualifier/Qualifier.cjs');
require('../../qualifiers/flag/FlagQualifier.cjs');
require('../../internal/utils/dataStructureUtils.cjs');
require('../../internal/models/ActionModel.cjs');
require('../../internal/models/actionToJson.cjs');
require('../../internal/utils/unsupportedError.cjs');
require('../../internal/models/QualifierModel.cjs');
require('../../internal/models/qualifierToJson.cjs');

/**
 * @description A class that defines how to remove the background of an asset
 * @extends SDK.Action
 * @memberOf Actions.Effect
 * @see Visit {@link Actions.Effect|Effect} for an example
 */
var BackgroundRemoval = /** @class */ (function (_super) {
    tslib_es6.__extends(BackgroundRemoval, _super);
    function BackgroundRemoval() {
        var _this = _super.call(this) || this;
        _this._actionModel.actionType = 'backgroundRemoval';
        return _this;
    }
    BackgroundRemoval.prototype.fineEdges = function (value) {
        this._fineEdges = value;
        if (this._fineEdges) {
            this._actionModel.fineEdges = this._fineEdges;
        }
        return this;
    };
    BackgroundRemoval.prototype.hints = function (value) {
        this._hints = value;
        if (this._hints) {
            this._actionModel.hints = this._hints;
        }
        return this;
    };
    BackgroundRemoval.prototype.prepareQualifiers = function () {
        var _a;
        var str = 'background_removal';
        if (this._fineEdges !== undefined) {
            str += ":" + new internal_qualifier_QualifierValue.QualifierValue("fineedges_" + (this._fineEdges ? 'y' : 'n')).toString();
        }
        if ((_a = this._hints) === null || _a === void 0 ? void 0 : _a.length) {
            str += ":" + new internal_qualifier_QualifierValue.QualifierValue("hints_(" + this._hints.join(';') + ")").toString();
        }
        this.addQualifier(new internal_qualifier_Qualifier.Qualifier('e', str));
    };
    BackgroundRemoval.fromJson = function (actionModel) {
        var _a = actionModel, fineEdges = _a.fineEdges, hints = _a.hints;
        // We are using this() to allow inheriting classes to use super.fromJson.apply(this, [actionModel])
        // This allows the inheriting classes to determine the class to be created
        return new this().fineEdges(fineEdges).hints(hints);
    };
    return BackgroundRemoval;
}(internal_Action.Action));

exports.BackgroundRemoval = BackgroundRemoval;
